### ISSUE TITLE
Keep uploaded file extension in link name

### ISSUE THREAD

**Issue by licaon-kter** [cmeng-git/atalk-android#66]
...so that Conversations shows its preview on screen and not just appear as a "Open file button"

Not sure it's based on extension or some MIME type.

2.1.1 F-Droid

**Comment by cmeng-git** [comment]
Actually the file name in the uploaded file is given by android OS when you pick a file from the android storage. aTalk does not change the file name; nor it knows what was the original file name if in fact android has changed it.

It seems that this is how android is designed to work, to prevent other app to have direct access to the file original location if it is not the rightful owner; giving the picker only a link without the mimeType extension.

Conversation uses mimeType extension in the file name to determine the type of file to show. Hence it has problem when it dealing a file without the mimeType extension.

aTalk has implemented the detection of the file memeType based on filename extension, and fallback to its content if formal detection failed.

**Comment by licaon-kter** [comment]
Can you rephrase your answer: "it doesn't work because...." ?

**Comment by cmeng-git** [comment]
In simple term, the filename in the uploaded file (and sent to the recipient) is given by android OS. aTalk does not known or way to determine the file original name.

**Comment by licaon-kter** [comment]
And when I attach in Conversations, Conversations can see the file EXTENSION because?

**Comment by cmeng-git** [comment]
May I known what android device/OS that you have install Conversation?

**Comment by licaon-kter** [comment]
Same device: Xiaomi MIUI11 Android 10

**Comment by cmeng-git** [comment]
That is strange, because aTalk will just use the filename given by the file picker as it, without changing it. Can you describe the steps how to reproduce the problem when you use aTalk to send file to Conversation.

Are you able to find out the original file name and the name used by aTalk when sending.

**Comment by licaon-kter** [comment]
OK, now I get it, when you "attach picture" you get a Recents view first, that I've used and got the button, hence this issue.

So now I tried again, ignored Recents, right side menu, selected Gallery -> Image, yielded a preview as expected.

But, using Recents to attach from Conversations still get the extension as expected, so I guess it needs some tuning on aTalk side.

**Comment by cmeng-git** [comment]
Thank,  I will take a look to understand the problem.

**Comment by cmeng-git** [comment]
I have checked the Conversations source, it does not use the original name from the file picker if android OS gives only the content:// scheme link; It make the best guess for the file mimeType, and then creates its own file name tagged with guessed mimeType extension.
Not sure if it is related to the best guess method; when I try to send a pdf file, Conversations crashes.

In aTalk implementation for OMEMO media sharing, it forwards the actual name returns by the file picker to the smack OMEMO for encryption. aTalk cannot modify the file name returns by android OS, otherwise smack will not be able to find the file for encryption. Conversations approach cannot be used in aTalk implementation.

aTalk implementation does not depends solely on extension to find the image file type. If you send files between two aTalk, aTalk is able to find the image file for view, even without the file extensions.

**Comment by cmeng-git** [comment]
Request declined - see above comments.
